extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -0,0 +1,479 @@
+            if (row != null) {
+              driver.add(row, committerSupplier);
+              metrics.incrementProcessed();
+            } else {
+              metrics.incrementUnparseable();
+            }
+
+            final long nextOffset = record.offset() + 1;
+            final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(record.partition());
+
+            nextOffsets.put(record.partition(), nextOffset);
+
+            if (nextOffset == endOffset && assignment.remove(record.partition())) {
+              log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
+              assignPartitions(consumer, topic, assignment);
+              stillReading = !assignment.isEmpty();
+            }
+          }
+        }
+      }
+
+      final TransactionalSegmentPublisher publisher = new TransactionalSegmentPublisher()
+      {
+        @Override
+        public boolean publishSegments(Set<DataSegment> segments, Object commitMetadata) throws IOException
+        {
+          // Sanity check, we should only be publishing things that match our desired end state.
+          if (!ioConfig.getEndPartitions().getPartitionOffsetMap().equals(commitMetadata)) {
+            throw new ISE("WTF?! Driver attempted to publish invalid metadata[%s].", commitMetadata);
+          }
+
+          final SegmentInsertAction action;
+
+          if (ioConfig.isUseTransaction()) {
+            action = new SegmentInsertAction(
+                segments,
+                new KafkaDataSourceMetadata(ioConfig.getStartPartitions()),
+                new KafkaDataSourceMetadata(ioConfig.getEndPartitions())
+            );
+          } else {
+            action = new SegmentInsertAction(segments, null, null);
+          }
+
+          log.info("Publishing with isTransaction[%s].", ioConfig.isUseTransaction());
+
+          return toolbox.getTaskActionClient().submit(action).isSuccess();
+        }
+      };
+
+      final SegmentsAndMetadata published = driver.finish(publisher, committerSupplier.get());
+      if (published == null) {
+        throw new ISE("Transaction failure publishing segments, aborting");
+      } else {
+        log.info(
+            "Published segments[%s] with metadata[%s].",
+            Joiner.on(", ").join(
+                Iterables.transform(
+                    published.getSegments(),
+                    new Function<DataSegment, String>()
+                    {
+                      @Override
+                      public String apply(DataSegment input)
+                      {
+                        return input.getIdentifier();
+                      }
+                    }
+                )
+            ),
+            published.getCommitMetadata()
+        );
+      }
+    }
+
+    return success();
+  }
+
+  @Override
+  public boolean canRestore()
+  {
+    // TODO
+    return false;
+  }
+
+  @Override
+  public void stopGracefully()
+  {
+    // TODO
+  }
+
```

Comment:
must be a restorable task and be able to stop gracefully. i know its todo :) , but just emphasizing.